cli/internal/helm/actionfactory.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package helm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/internal/compatibility"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/kubernetes/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/semver"
+	"github.com/spf13/afero"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// ErrConfirmationMissing signals that an action requires user confirmation.
+var ErrConfirmationMissing = errors.New("action requires user confirmation")
+
+var errReleaseNotFound = errors.New("release not found")
+
+type actionFactory struct {
+	versionLister ReleaseVersionLister
+	cfg           *action.Configuration
+	kubeClient    crdClient
+	log           debugLog
+}
+
+type crdClient interface {
+	Initialize(kubeconfig []byte) error
+	ApplyCRD(ctx context.Context, rawCRD []byte) error
+	GetCRDs(ctx context.Context) ([]apiextensionsv1.CustomResourceDefinition, error)
+	GetCRs(ctx context.Context, gvr schema.GroupVersionResource) ([]unstructured.Unstructured, error)
+}
+
+// newActionFactory creates a new action factory for managing helm releases.
+func newActionFactory(kubeConfigPath string, log debugLog) (*actionFactory, error) {
```

Comment:
to handle potential migration actions for charts we could add
`Register(releaseName,preHook,postHook)` to the factory or even just as a function with `[]applyActions` and set the respective hook on the action.